### PR TITLE
Support deploy TiFlash on multi-disks with "storage" configurations since v4.0.9

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve? 

[enhancement for TiFlash multi-disks deployment](https://docs.google.com/document/d/1QzgqqNimOxvRMuaGDmTQX83M0mlk-mWjbCUZM1PgTSY/edit)
TiFlash has added some new configurations (storage.*) to improve performance under multi-disk deployment, since v4.0.9.

This PR
* supports new configurations (storage.*)
* validates whether the path configurations are valid for deploying/edit-config/scale-out
* keeps backward compatibility for deploying cluster with version less than v4.0.9

### What is changed and how it works?
* Rewrite TiFlashSpec DataDir in unmarshal with overridden dirs
* When users want to deploy/scal-out/edit-config, `Specification.Validate` will check whether the configuration in TiFlash spec is valid or not
* Mark TiFlashSpec DataDir is expandable so that allows users to expand the data directories of TiFlash. (Shrinking directories is not allowed)
* Refine the generated configuration file for version >= v4.0.9, and also keep backward compatibility for version < v4.0.9

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
  [TiUP test cases for TiFlash multi-disks](https://docs.google.com/document/d/1KlyLjwEg9tRlM5Hcb38J35zVFS5N67RhoU9ga3VxqPQ/edit#)

Code changes

 - Has exported function/method change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation (TODO)

Release notes:

```release-note
Support deploy TiFlash node with `storage` configurations
```

### CODE REVIEW

File: pkg/cluster/spec/spec_test.go [hunk 1]
```diff
@@ -309,7 +310,7 @@ item2 = 500
 item5 = 700
 item6 = 600
 `
-	got, err := merge2Toml("tikv", topo.ServerConfigs.TiKV, topo.TiKVServers[0].Config)
+	got, err := merge2Toml(ComponentTiKV, topo.ServerConfigs.TiKV, topo.TiKVServers[0].Config)
```
Comment (AstroProfundis) [inline pkg/cluster/spec/spec_test.go:313]: I believe it's mean to be hard coded.

**Review by AstroProfundis — COMMENTED** [review]
I don't think it's a good idea for TiUP Cluster to be aware of what `config` of a server node is, it's designed to be an interface by intend.

Maybe we can reuse current specification of `data_dir` and use the first item in the list as `main.dir`, so that the specification of other components could remain unchanged.

**Comment by JaySon-Huang** [comment]
> Maybe we can reuse current specification of data_dir and use the first item in the list as main.dir, so that the specification of other components could remain unchanged.

@AstroProfundis The current specification of `data_dir` can not exactly represent the "storage.*" configurations when users 
 want to deploy on a multi-disk node.

```
data_dir: "/data0/tiflash,/data1/tiflash,/data2/tiflash"
config:
  path_realtime_mode = false
```
is equal to:
```
config:
  storage.main.dir: ["/data0/tiflas", "/data1/tiflash", "/data2/tiflash"]
  storage.latest.dir: ["/data0/tiflash"]
```
But if users want to fully make use of their disks IO after v4.0.9, they need to change their configs like this
```
config:
  storage.main.dir: ["/data0/tiflas", "/data1/tiflash", "/data2/tiflash"]
  storage.latest.dir: ["/data0/tiflash", "/data1/tiflash", "/data2/tiflash"]
```

**Comment by JaySon-Huang** [comment]
These configurations are related to the directories to store data. We need to detect conflict directories when deploying, remove them after TiFlash gets destroyed. And validate them when they are changed.

**Comment by AstroProfundis** [comment]
OK that sounds reasonable to me

**Comment by JaySon-Huang** [comment]
After a cluster destroyed, the directories defined in "storage.*" didn't get destroyed.
It is because that some methods (like `Specification::CountDir` in validate.go) use reflection and get the wrong directories from `TiFlashSpec::DataDir` but not `TiFlashSpec::GetDataDir`.

Do you have any idea to solve this problem? @AstroProfundis @lucklove

File: pkg/cluster/spec/tiflash.go [hunk 2]
```diff
@@ -87,6 +86,68 @@ func (s TiFlashSpec) IsImported() bool {
...
+		return strs
+	}
+	mainDirs := getStrings(TiFlashStorageKeyMainDirs)
+	latestDirs := getStrings(TiFlashStorageKeyLatestDirs)
+	if len(mainDirs) == 0 && len(latestDirs) == 0 {
+		return s.DataDir
+	}
+	// If storage is defined, the path defined in "data_dir" will be ignored
+	// make the dirs uniq
+	dirSet := make(map[string]bool)
```
[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply pkg/cluster/spec/tiflash.go]: addressed

**Comment by lucklove** [comment]
> After a cluster destroyed, the directories defined in "storage.*" didn't get destroyed.
> It is because that some methods (like `Specification::CountDir` in validate.go) use reflection and get the wrong directories from `TiFlashSpec::DataDir` but not `TiFlashSpec::GetDataDir`.
> 
> Do you have any idea to solve this problem? @AstroProfundis @lucklove

IMO, we shouldn't add the function `GetDataDir`, but reset the DataDir filed to the correct one in https://github.com/pingcap/tiup/blob/master/pkg/cluster/spec/spec.go#L279. Because there are many places use reflect to deal with DeployDir, DataDir and LogDir

**Comment by JaySon-Huang** [comment]
Have added a document about manual test cases and result document in the PR description.
PTAL again @lucklove @AstroProfundis

**Review by lucklove — APPROVED** [review]
LGTM

**Pull request merged**